src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -465,3 +598,207 @@ ShardPlacementForFunctionColocatedWithReferenceTable(CitusTableCacheEntry *cache
+	{
+		return NULL;
+	}
+
+	if (list_length(rtentry->functions) != 1)
+	{
+		/* Skip if RTE isn't a single FuncExpr */
+		return NULL;
+	}
+
+	RangeTblFunction *rtfunc = (RangeTblFunction *) linitial(rtentry->functions);
+	if (!IsA(rtfunc->funcexpr, FuncExpr))
+	{
+		/* Skip if RTE isn't a simple FuncExpr */
+		return NULL;
+	}
+
+	return (FuncExpr *) rtfunc->funcexpr;
+}
+
+
+/*
+ * Sets a flag to true indicating that the current node is executing a delegated
+ * function call, using forceDelegation, within a distributed transaction issued
+ * by the coordinator. Also, saves the distribution argument.
+ */
+static void
+EnableInForceDelegatedFuncExecution(Const *distArgument, uint32 colocationId)
+{
+	/*
+	 * The saved distribution argument need to persist through the life
+	 * of the query, both during the planning (where we save) and execution
+	 * (where we compare)
+	 */
+	MemoryContext oldcontext = MemoryContextSwitchTo(TopTransactionContext);
+	ereport(DEBUG1, errmsg("Saving Distribution Argument: %s:%d",
+						   pretty_format_node_dump(nodeToString(distArgument)),
+						   colocationId));
+	AllowedDistributionColumnValue.distributionColumnValue = copyObject(distArgument);
+	AllowedDistributionColumnValue.colocationId = colocationId;
+	AllowedDistributionColumnValue.isActive = true;
+	MemoryContextSwitchTo(oldcontext);
+}
+
+
+/*
+ * Within a 2PC, when a function is delegated to a remote node, we pin
+ * the distribution argument as the shard key for all the SQL in the
+ * function's block. The restriction is imposed to not to access other
+ * nodes from the current node and violate the transactional integrity of
+ * the 2PC. Reset the distribution argument value once the function ends.
+ */
+void
+ResetAllowedShardKeyValue(void)
+{
+	if (AllowedDistributionColumnValue.isActive)
+	{
+		pfree(AllowedDistributionColumnValue.distributionColumnValue);
+		AllowedDistributionColumnValue.isActive = false;
+	}
+
+	InTopLevelDelegatedFunctionCall = false;
+}
+
+
+/*
+ * Function returns true if the current shard key in the adaptive executor
+ * matches the saved distribution argument of a force_delegation function.
+ */
+bool
+IsShardKeyValueAllowed(Const *shardKey, uint32 colocationId)
+{
+	Assert(AllowedDistributionColumnValue.isActive);
+
+	ereport(DEBUG4, errmsg("Comparing saved:%s with Shard key: %s colocationid:%d:%d",
+						   pretty_format_node_dump(
+							   nodeToString(
+								   AllowedDistributionColumnValue.distributionColumnValue)),
+						   pretty_format_node_dump(nodeToString(shardKey)),
+						   AllowedDistributionColumnValue.colocationId, colocationId));
+
+	return (equal(AllowedDistributionColumnValue.distributionColumnValue, shardKey) &&
+			(AllowedDistributionColumnValue.colocationId == colocationId));
+}
+
+
+/*
+ * Sets the flag InTopLevelDelegatedFunctionCall to true if the current
+ * Citus backend is allowed to do remote tasks.
+ */
+static void
+CheckRemoteTaskExecutionAllowed(bool IsForceDelegated)
```

Comment:
IsForceDelegated -> isForceDelegated

we usually only start with a capital for global variables

Reply:
Changed to isForceDelegated